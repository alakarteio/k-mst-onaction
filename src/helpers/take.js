import { isRegExp, isString, isFunction } from 'lodash'

const take = (match, callback) => (action, tree) => {
  const isMatching = false
    || (isString(match) && match === action.fullpath)
    || (isFunction(match) && match(action, tree))
    || (isRegExp(match) && action.fullpath.match(match))

  if (isMatching) return callback(action, tree)
  return false
}

const ended = (match, callback) => (action, tree) => {
  if (action.ended) return take(match, callback)(action, tree)
  return false
}

take.ended = ended

export default take
