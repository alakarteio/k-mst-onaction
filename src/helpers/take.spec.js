/* eslint-env jest */
import take from './take'

describe('take helper', () => {
  describe('simple version', () => {
    it('should not match (wrong type)', () => {
      const callback = jest.fn(() => 'callback returns')

      const res = take(['do', 'not', 'match', 'this', 'please'], callback)({}, {})

      expect(res).toMatchSnapshot()
      expect(callback.mock.calls).toMatchSnapshot()
    })

    it('should match -string-', () => {
      const callback = jest.fn(() => 'callback returns')
      const fullpath = '/some/full/path'

      const res = take(fullpath, callback)({ fullpath }, { this: 'is tree' })

      expect(res).toMatchSnapshot()
      expect(callback.mock.calls).toMatchSnapshot()
    })
    it('should **not** match -string-', () => {
      const callback = jest.fn(() => 'callback returns')
      const fullpath = '/some/full/path'

      const res = take('oups', callback)({ fullpath }, { this: 'is tree' })

      expect(res).toMatchSnapshot()
      expect(callback.mock.calls).toMatchSnapshot()
    })

    // ==========
    // from now tests are simpler because both returns type are tested
    // ==========

    it('should match -function-', () => {
      const res = take(() => true, () => 'match !')({}, {})
      expect(res).toMatchSnapshot()
    })
    it('should **not** match -function-', () => {
      const res = take(() => false, () => 'match !')({}, {})
      expect(res).toMatchSnapshot()
    })

    it('should match -regexp-', () => {
      const res = take(/setText/, () => 'match !')({ fullpath: '/yeah/path/to/setText' }, {})
      expect(res).toMatchSnapshot()
    })
    it('should **not** match -regexp-', () => {
      const res = take(/settext/, () => 'match !')({ fullpath: '/yeah/path/to/setText' }, {})
      expect(res).toMatchSnapshot()
    })
  })

  describe('ended version', () => {
    // here we test only one type because other are tested before
    it('should match -ended: true-', () => {
      const callback = jest.fn(() => 'callback returns !')
      const fullpath = 'match please'
      const res = take.ended(fullpath, callback)({ fullpath, ended: true }, { this: 'is tree :)' })

      expect(res).toMatchSnapshot()
      expect(callback.mock.calls).toMatchSnapshot()
    })

    it('should not match -ended: false-', () => {
      const callback = jest.fn(() => 'callback returns !')
      const fullpath = 'match please'
      const res = take.ended(fullpath, callback)({ fullpath }, { this: 'is tree :)' })

      expect(res).toMatchSnapshot()
      expect(callback.mock.calls).toMatchSnapshot()
    })
  })
})
