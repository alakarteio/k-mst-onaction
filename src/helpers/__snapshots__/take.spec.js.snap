// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`take helper ended version should match -ended: true- 1`] = `"callback returns !"`;

exports[`take helper ended version should match -ended: true- 2`] = `
Array [
  Array [
    Object {
      "ended": true,
      "fullpath": "match please",
    },
    Object {
      "this": "is tree :)",
    },
  ],
]
`;

exports[`take helper ended version should not match -ended: false- 1`] = `false`;

exports[`take helper ended version should not match -ended: false- 2`] = `Array []`;

exports[`take helper simple version should **not** match -function- 1`] = `false`;

exports[`take helper simple version should **not** match -regexp as string- 1`] = `false`;

exports[`take helper simple version should **not** match -regexp- 1`] = `false`;

exports[`take helper simple version should **not** match -string- 1`] = `false`;

exports[`take helper simple version should **not** match -string- 2`] = `Array []`;

exports[`take helper simple version should match -function- 1`] = `"match !"`;

exports[`take helper simple version should match -regexp as string- 1`] = `"match !"`;

exports[`take helper simple version should match -regexp- 1`] = `"match !"`;

exports[`take helper simple version should match -string- 1`] = `"callback returns"`;

exports[`take helper simple version should match -string- 2`] = `
Array [
  Array [
    Object {
      "fullpath": "/some/full/path",
    },
    Object {
      "this": "is tree",
    },
  ],
]
`;

exports[`take helper simple version should not match (wrong type) 1`] = `false`;

exports[`take helper simple version should not match (wrong type) 2`] = `Array []`;
